File: Reqnroll/Chapter15/15.7-TheStubDependencyPattern/After/WIMP.Specs/Support/LambdaAction.cs
namespace WIMP.Specs.Support;

public class LambdaAction<TResult>(string testActionName, Func<Task<TResult>> action)
    : TestAction<TResult>(testActionName)
{
    /// <summary>
    /// Creates a lambda action with a synchronous action
    /// </summary>
    public LambdaAction(string testActionName, Func<TResult> action)
        : this(testActionName, () => Task.FromResult(action()))
    {
    }

    protected override async Task<TResult> DoExecute()
    {
        return await action();
    }
}

/// <summary>
/// Lambda action for void-return actions
/// </summary>
public class LambdaAction(string testActionName, Func<Task> action) :
    LambdaAction<VoidReturn>(testActionName,
        () => action().ContinueWith(_ => VoidReturn.Instance))
{
    /// <summary>
    /// Creates a void-return lambda action with a synchronous action
    /// </summary>
    public LambdaAction(string testActionName, Action action)
        : this(testActionName, () =>
        {
            action();
            return Task.CompletedTask;
        })
    {
    }
}

File: CucumberJVM/Chapter15/15.7-StubDependencyPattern/After/src/test/java/com/wimp/specs/support/LambdaAction.java
package com.wimp.specs.support;

public class LambdaAction<T> implements TestAction<T> {
    @FunctionalInterface
    public interface ThrowingSupplier<TValue> {
        TValue get() throws Exception;
    }

    private final ThrowingSupplier<T> action;

    public LambdaAction(ThrowingSupplier<T> action) {
        this.action = action;
    }

    @Override
    public T execute() throws Exception {
        return action.get();
    }
}
